File: strm_sync.py
"""
STRM 同步模块
功能：
1. 根据库文件配置，自动同步 STRM 文件和字幕文件
2. 自动删除多余的 STRM 文件和字幕文件
3. 支持全量同步和增量同步
"""
import json
import logging
from pathlib import Path
from typing import Dict, List, Tuple, Optional, Set
from concurrent.futures import ThreadPoolExecutor, as_completed
import threading

from strm_app import (
    VIDEO_EXTS,
    SUBTITLE_EXTS,
    CATEGORY_DIRS,
    safe_rel_path,
    make_play_url,
    download_subtitle_file,
    load_lib,
    DEFAULT_OUTPUT_DIR,
)

logger = logging.getLogger(__name__)

# 同步状态
SYNC_STATUS = {}
SYNC_LOCK = threading.Lock()


def _cat_root(output_dir: str, category: str) -> Path:
    """计算库文件实际输出根目录（含分类子目录）"""
    root = Path(output_dir).expanduser()
    if category and category in CATEGORY_DIRS:
        root = root / CATEGORY_DIRS[category]
    return root


def get_expected_files(lib: Dict, output_dir: str, include_subtitles: bool = False) -> Tuple[List[Path], List[Path]]:
    """
    根据库文件配置，计算期望生成的 STRM 文件和字幕文件列表

    Returns:
        (strm_files, subtitle_files)
    """
    category = lib.get('category', '')
    out_root = _cat_root(output_dir, category)

    strm_files = []
    subtitle_files = []

    for f in lib.get('files', []):
        path = f.get('path', '')
        if not path:
            continue
        rel = safe_rel_path(path)
        ext = rel.suffix.lower()
        if ext in VIDEO_EXTS:
            strm_files.append(out_root / rel.with_suffix('.strm'))
        elif include_subtitles and ext in SUBTITLE_EXTS:
            subtitle_files.append(out_root / rel)

    return strm_files, subtitle_files


def get_existing_files(scan_root: Path) -> Tuple[List[Path], List[Path]]:
    """
    获取输出目录中现有的 STRM 文件和字幕文件

    Returns:
        (strm_files, subtitle_files)
    """
    if not scan_root.exists():
        return [], []

    strm_files = list(scan_root.rglob('*.strm'))
    subtitle_files = [f for f in scan_root.rglob('*') if f.suffix.lower() in SUBTITLE_EXTS]
    return strm_files, subtitle_files


def _collect_expected(libs: List[Dict], output_dir: str, include_subtitles: bool) -> Tuple[Set[Path], Set[Path], Dict[Path, Dict]]:
    """合并多个库的期望文件，并构建「目标路径 -> 文件信息」映射"""
    strm_map: Dict[Path, Dict] = {}
    sub_map: Dict[Path, Dict] = {}
    for lib in libs:
        category = lib.get('category', '')
        cat_root = _cat_root(output_dir, category)
        for f in lib.get('files', []):
            path = f.get('path', '')
            if not path:
                continue
            rel = safe_rel_path(path)
            ext = rel.suffix.lower()
            if ext in VIDEO_EXTS:
                strm_map.setdefault(cat_root / rel.with_suffix('.strm'), f)
            elif include_subtitles and ext in SUBTITLE_EXTS:
                sub_map.setdefault(cat_root / rel, f)
    return set(strm_map.keys()), set(sub_map.keys()), strm_map, sub_map


def _sync(
    expected_strm: Set[Path],
    expected_subs: Set[Path],
    strm_map: Dict[Path, Dict],
    sub_map: Dict[Path, Dict],
    output_dir: str,
    server_base: str,
    include_subtitles: bool,
    dry_run: bool,
    label: str = '',
    scan_root: Path = None,
) -> Dict:
    """核心同步：对比期望与现有，删除多余、生成缺失"""
    out_root_dir = Path(output_dir).expanduser()
    if scan_root is None:
        scan_root = out_root_dir
    existing_strm, existing_subs = get_existing_files(scan_root)
    existing_strm_set = set(existing_strm)
    existing_subs_set = set(existing_subs)

    to_delete_strm = existing_strm_set - expected_strm
    to_delete_subs = existing_subs_set - expected_subs
    to_create_strm = expected_strm - existing_strm_set
    to_create_subs = expected_subs - existing_subs_set

    result = {
        'lib_id': label,
        'expected_strm': len(expected_strm),
        'expected_subs': len(expected_subs),
        'existing_strm': len(existing_strm),
        'existing_subs': len(existing_subs),
        'to_delete_strm': len(to_delete_strm),
        'to_delete_subs': len(to_delete_subs),
        'to_create_strm': len(to_create_strm),
        'to_create_subs': len(to_create_subs),
        'deleted_strm': [],
        'deleted_subs': [],
        'created_strm': [],
        'created_subs': [],
        'errors': [],
    }

    if dry_run:
        result['dry_run'] = True
        logger.info(f"[干跑] {label}: 期望 {len(expected_strm)} 个 STRM, 现有 {len(existing_strm)} 个")
        logger.info(f"[干跑] 需删除: {len(to_delete_strm)} 个 STRM, {len(to_delete_subs)} 个字幕")
        logger.info(f"[干跑] 需生成: {len(to_create_strm)} 个 STRM, {len(to_create_subs)} 个字幕")
        return result

    rel_of = lambda p: str(p.relative_to(out_root_dir))

    # 删除多余的 STRM 文件
    for f in to_delete_strm:
        try:
            f.unlink()
            result['deleted_strm'].append(rel_of(f))
            logger.info(f"删除 STRM: {f}")
        except Exception as e:
            result['errors'].append(f"删除 STRM 失败 {f}: {e}")
            logger.error(f"删除 STRM 失败 {f}: {e}")

    # 删除多余的字幕文件
    for f in to_delete_subs:
        try:
            f.unlink()
            result['deleted_subs'].append(rel_of(f))
            logger.info(f"删除字幕: {f}")
        except Exception as e:
            result['errors'].append(f"删除字幕失败 {f}: {e}")
            logger.error(f"删除字幕失败 {f}: {e}")

    # 清理空目录（仅当目录为空时）
    def cleanup_empty_dirs(base_dir: Path, stop_at: Path = None):
        if base_dir == stop_at or not base_dir.exists():
            return
        try:
            for subdir in list(base_dir.iterdir()):
                if subdir.is_dir():
                    cleanup_empty_dirs(subdir, stop_at)
            if not any(base_dir.iterdir()):
                base_dir.rmdir()
                logger.info(f"清理空目录: {base_dir}")
        except Exception:
            pass

    cleanup_empty_dirs(scan_root, stop_at=out_root_dir if scan_root != out_root_dir else None)

    # 生成缺失的 STRM 文件
    def create_strm(strm_path: Path):
        file_info = strm_map.get(strm_path)
        if file_info is None:
            return False
        url = make_play_url(server_base, file_info['idx'], file_info['etag'],
                            int(file_info.get('size') or 0), Path(file_info['path']).name)
        strm_path.parent.mkdir(parents=True, exist_ok=True)
        strm_path.write_text(url + '\n', encoding='utf-8')
        result['created_strm'].append(rel_of(strm_path))
        logger.info(f"生成 STRM: {strm_path}")
        return True

    with ThreadPoolExecutor(max_workers=8) as executor:
        futures = {executor.submit(create_strm, f): f for f in to_create_strm}
        for future in as_completed(futures):
            strm_path = futures[future]
            try:
                future.result()
            except Exception as e:
                result['errors'].append(f"生成 STRM 失败 {strm_path}: {e}")
                logger.error(f"生成 STRM 失败 {strm_path}: {e}")

    # 下载缺失的字幕文件
    if include_subtitles:
        cfg = {}
        try:
            from strm_app import config
            cfg = config()
        except Exception:
            pass

        # 未登录 123 网盘时跳过字幕下载，避免逐个尝试登录导致卡顿
        try:
            from strm_app import CACHE_PATH
            import time as _time
            _cached = {}
            try:
                _cached = json.loads(CACHE_PATH.read_text(encoding='utf-8'))
            except Exception:
                pass
            logged_in = (_cached.get('accessToken') and _cached.get('tokenCreateTime')
                         and _time.time() - int(_cached.get('tokenCreateTime') or 0) < 25 * 24 * 60 * 60)
        except Exception:
            logged_in = False

        if not logged_in and to_create_subs:
            result['subtitle_skipped'] = '未登录 123 网盘，跳过字幕下载'
            logger.warning("未登录 123 网盘，跳过字幕下载")
            return result

        def download_sub(sub_path: Path):
            file_info = sub_map.get(sub_path)
            if file_info is None:
                return False
            try:
                fast_mode = cfg.get('mode', 'cache') == 'fast'
                if download_subtitle_file(file_info, sub_path, fast_mode):
                    result['created_subs'].append(rel_of(sub_path))
                    logger.info(f"下载字幕: {sub_path}")
                    return True
            except Exception as e:
                logger.error(f"下载字幕失败 {sub_path}: {e}")
            return False

        with ThreadPoolExecutor(max_workers=8) as executor:
            futures = {executor.submit(download_sub, f): f for f in to_create_subs}
            for future in as_completed(futures):
                sub_path = futures[future]
                try:
                    future.result()
                except Exception as e:
                    result['errors'].append(f"下载字幕失败 {sub_path}: {e}")
                    logger.error(f"下载字幕失败 {sub_path}: {e}")

    return result


def sync_all_libraries(
    output_dir: str = None,
    server_base: str = None,
    include_subtitles: bool = False,
    dry_run: bool = False,
) -> Dict:
    """
    同步所有库的 STRM 文件：合并全部库的期望后整体对比，避免同分类多库互相误删。

    Returns:
        总体同步结果
    """
    from strm_app import config, list_libraries

    cfg = config()
    if not output_dir:
        output_dir = cfg.get('output_dir') or DEFAULT_OUTPUT_DIR
    if not server_base:
        server_base = cfg.get('server_base') or 'http://127.0.0.1:8000'

    out_root_dir = Path(output_dir).expanduser()

    libraries = list_libraries()
    all_libs = [load_lib(info['id']) for info in libraries if info.get('id')]

    expected_strm, expected_subs, strm_map, sub_map = _collect_expected(all_libs, output_dir, include_subtitles)

    result = _sync(expected_strm, expected_subs, strm_map, sub_map, output_dir, server_base,
                   include_subtitles, dry_run, label='all',
                   scan_root=out_root_dir)

    total_result = {
        'libraries': result,
        'total_expected_strm': result['expected_strm'],
        'total_existing_strm': result['existing_strm'],
        'total_to_delete_strm': result['to_delete_strm'],
        'total_to_create_strm': result['to_create_strm'],
        'total_deleted_strm': len(result['deleted_strm']),
        'total_created_strm': len(result['created_strm']),
        'total_deleted_subs': len(result['deleted_subs']),
        'total_created_subs': len(result['created_subs']),
        'total_errors': len(result['errors']),
    }
    return total_result

File: internal/app/strm_sync.go
package app

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

func (a *App) catRoot(outputDir, category string) string {
	root := filepath.Clean(outputDir)
	if sub, ok := CATEGORY_DIRS[category]; ok {
		root = filepath.Join(root, sub)
	}
	return root
}

// getExpectedFiles: 计算期望生成的 STRM 和附属文件（字幕/nfo/图片）
func (a *App) getExpectedFiles(lib map[string]any, outputDir string, ds downloadSettings) ([]string, []string) {
	category, _ := lib["category"].(string)
	outRoot := a.catRoot(outputDir, category)
	var strmFiles, sideFiles []string
	files, _ := lib["files"].([]any)
	for _, f := range files {
		fm, ok := f.(map[string]any)
		if !ok {
			continue
		}
		rel := safeRelPath(asString(fm["path"]))
		ext := strings.ToLower(filepath.Ext(rel))
		if VIDEO_EXTS[ext] {
			strmFiles = append(strmFiles, filepath.Join(outRoot, relWithoutSuffix(rel, ext)+".strm"))
		} else if kind := sidecarType(ext); kind != "" && ds.wants(kind) {
			sideFiles = append(sideFiles, filepath.Join(outRoot, rel))
		}
	}
	return strmFiles, sideFiles
}

// getExistingFiles: 扫描输出目录中已有的 STRM 与需要管理的附属文件。
// 只统计“当前启用类型”的附属文件，避免误删用户手动放置的其它元数据。
func getExistingFiles(scanRoot string, ds downloadSettings) ([]string, []string) {
	if _, err := os.Stat(scanRoot); err != nil {
		return nil, nil
	}
	var strmFiles, sideFiles []string
	filepath.Walk(scanRoot, func(p string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		if strings.EqualFold(filepath.Ext(p), ".strm") {
			strmFiles = append(strmFiles, p)
		} else if kind := sidecarType(strings.ToLower(filepath.Ext(p))); kind != "" && ds.wants(kind) {
			sideFiles = append(sideFiles, p)
		}
		return nil
	})
	return strmFiles, sideFiles
}

func collectExpected(libs []map[string]any, outputDir string, ds downloadSettings) (map[string]bool, map[string]bool, map[string]map[string]any, map[string]map[string]any) {
	strmMap := map[string]map[string]any{}
	sideMap := map[string]map[string]any{}
	for _, lib := range libs {
		category, _ := lib["category"].(string)
		catRoot := a_catRoot(outputDir, category)
		files, _ := lib["files"].([]any)
		for _, f := range files {
			fm, ok := f.(map[string]any)
			if !ok {
				continue
			}
			rel := safeRelPath(asString(fm["path"]))
			ext := strings.ToLower(filepath.Ext(rel))
			if VIDEO_EXTS[ext] {
				key := filepath.Join(catRoot, relWithoutSuffix(rel, ext)+".strm")
				if _, exists := strmMap[key]; !exists {
					strmMap[key] = fm
				}
			} else if kind := sidecarType(ext); kind != "" && ds.wants(kind) {
				key := filepath.Join(catRoot, rel)
				if _, exists := sideMap[key]; !exists {
					sideMap[key] = fm
				}
			}
		}
	}
	strmSet := map[string]bool{}
	for k := range strmMap {
		strmSet[k] = true
	}
	sideSet := map[string]bool{}
	for k := range sideMap {
		sideSet[k] = true
	}
	return strmSet, sideSet, strmMap, sideMap
}

func a_catRoot(outputDir, category string) string {
	root := filepath.Clean(outputDir)
	if sub, ok := CATEGORY_DIRS[category]; ok {
		root = filepath.Join(root, sub)
	}
	return root
}

// syncCore: 核心同步：对比期望与现有，删除多余、生成缺失
func (a *App) syncCore(expectedStrm, expectedSide map[string]bool, strmMap, sideMap map[string]map[string]any, outputDir, serverBase string, ds downloadSettings, dryRun bool, label string, scanRoot string) map[string]any {
	outRootDir := filepath.Clean(outputDir)
	if scanRoot == "" {
		scanRoot = outRootDir
	}
	existingStrm, existingSide := getExistingFiles(scanRoot, ds)
	existingStrmSet := map[string]bool{}
	for _, f := range existingStrm {
		existingStrmSet[f] = true
	}
	existingSideSet := map[string]bool{}
	for _, f := range existingSide {
		existingSideSet[f] = true
	}

	toDeleteStrm := setDiff(existingStrmSet, expectedStrm)
	toDeleteSide := setDiff(existingSideSet, expectedSide)
	toCreateStrm := setDiff(expectedStrm, existingStrmSet)
	toCreateSide := setDiff(expectedSide, existingSideSet)

	result := map[string]any{
		"lib_id":         label,
		"expected_strm":  len(expectedStrm),
		"expected_subs":  len(expectedSide),
		"existing_strm":  len(existingStrm),
		"existing_subs":  len(existingSide),
		"to_delete_strm": len(toDeleteStrm),
		"to_delete_subs": len(toDeleteSide),
		"to_create_strm": len(toCreateStrm),
		"to_create_subs": len(toCreateSide),
		"deleted_strm":   []string{},
		"deleted_subs":   []string{},
		"created_strm":   []string{},
		"created_subs":   []string{},
		"errors":         []string{},
	}

	if dryRun {
		result["dry_run"] = true
		return result
	}

	relOf := func(p string) string {
		r, err := filepath.Rel(outRootDir, p)
		if err != nil {
			return p
		}
		return r
	}

	// 删除多余的 STRM
	for f := range toDeleteStrm {
		if err := os.Remove(f); err != nil {
			result["errors"] = append(result["errors"].([]string), "删除 STRM 失败 "+f+": "+err.Error())
		} else {
			result["deleted_strm"] = append(result["deleted_strm"].([]string), relOf(f))
			log.Printf("[同步] 删除 STRM: %s", relOf(f))
		}
	}

	// 删除多余的附属文件（字幕/nfo/图片）
	for f := range toDeleteSide {
		if err := os.Remove(f); err != nil {
			result["errors"] = append(result["errors"].([]string), "删除附属文件失败 "+f+": "+err.Error())
		} else {
			result["deleted_subs"] = append(result["deleted_subs"].([]string), relOf(f))
			log.Printf("[同步] 删除附属文件: %s", relOf(f))
		}
	}

	// 清理空目录
	a.cleanupEmptyDirs(scanRoot, outRootDir)

	// 生成缺失的 STRM（并行 8）
	var mu sync.Mutex // 保护 result map 的并发写入
	createStrm := func(strmPath string) {
		fileInfo, ok := strmMap[strmPath]
		if !ok {
			return
		}
		fileName := filepath.Base(safeRelPath(asString(fileInfo["path"])))
		url := makePlayURL(serverBase, int(firstInt64(fileInfo, "idx")), asString(fileInfo["etag"]), firstInt64(fileInfo, "size"), fileName)
		os.MkdirAll(filepath.Dir(strmPath), 0o755)
		os.WriteFile(strmPath, []byte(url+"\n"), 0o644)
		mu.Lock()
		result["created_strm"] = append(result["created_strm"].([]string), relOf(strmPath))
		mu.Unlock()
	}
	parallelStrings(toCreateStrm, 8, func(s string) {
		if err := safeRun(func() { createStrm(s) }); err != nil {
			mu.Lock()
			result["errors"] = append(result["errors"].([]string), "生成 STRM 失败 "+s+": "+err.Error())
			mu.Unlock()
		}
	})

	// 下载缺失的附属文件（字幕/nfo/图片），线程数与重试次数由配置控制
	if ds.enabled && len(toCreateSide) > 0 {
		cfg := a.cfg.Config()
		loggedIn := a.panLoggedIn()
		if !loggedIn {
			result["sidecar_skipped"] = "未登录 123 网盘，跳过附属文件下载"
			return result
		}
		fastMode := cfg["mode"] == "fast"
		downloadSide := func(path string) {
			fileInfo, ok := sideMap[path]
			if !ok {
				return
			}
			if a.downloadSidecarFile(fileInfo, path, asBool(fastMode), ds.retries) {
				mu.Lock()
				result["created_subs"] = append(result["created_subs"].([]string), relOf(path))
				mu.Unlock()
			}
		}
		parallelStrings(toCreateSide, ds.threads, func(s string) {
			if err := safeRun(func() { downloadSide(s) }); err != nil {
				mu.Lock()
				result["errors"] = append(result["errors"].([]string), "下载附属文件失败 "+s+": "+err.Error())
				mu.Unlock()
			}
		})
	}
	return result
}

func asBool(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	case string:
		return t == "true" || t == "1" || t == "fast"
	}
	return false
}

func (a *App) panLoggedIn() bool {
	cacheData := ReadJSONFile(a.cfg.CachePath)
	if cacheData == nil {
		return false
	}
	tok, _ := cacheData["accessToken"].(string)
	if tok == "" {
		return false
	}
	ct, ok := cacheData["tokenCreateTime"]
	if !ok {
		return false
	}
	var t int64
	switch v := ct.(type) {
	case float64:
		t = int64(v)
	case int64:
		t = v
	case string:
		t = parseInt(v)
	}
	return time.Now().Unix()-t < 25*24*60*60
}

func (a *App) cleanupEmptyDirs(baseDir, stopAt string) {
	if filepath.Clean(baseDir) == filepath.Clean(stopAt) {
		return
	}
	entries, err := os.ReadDir(baseDir)
	if err != nil {
		return
	}
	for _, e := range entries {
		if e.IsDir() {
			a.cleanupEmptyDirs(filepath.Join(baseDir, e.Name()), stopAt)
		}
	}
	entries2, _ := os.ReadDir(baseDir)
	if len(entries2) == 0 && baseDir != stopAt {
		os.Remove(baseDir)
	}
}

func setDiff(a, b map[string]bool) map[string]bool {
	out := map[string]bool{}
	for k := range a {
		if !b[k] {
			out[k] = true
		}
	}
	return out
}

func safeRun(f func()) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmtErr(r)
		}
	}()
	f()
	return nil
}

func fmtErr(r any) error {
	if e, ok := r.(error); ok {
		return e
	}
	return errUnsupportedFormat
}

func parallelStrings(set map[string]bool, workers int, fn func(string)) {
	keys := make([]string, 0, len(set))
	for k := range set {
		keys = append(keys, k)
	}
	ch := make(chan string)
	var wg sync.WaitGroup
	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for s := range ch {
				fn(s)
			}
		}()
	}
	for _, k := range keys {
		ch <- k
	}
	close(ch)
	wg.Wait()
}

// syncAllLibraries: 合并全部库的期望后整体对比
func (a *App) syncAllLibraries(outputDir, serverBase string, includeSubtitles bool) map[string]any {
	cfg := a.cfg.Config()
	ds := a.getDownloadSettings()
	if outputDir == "" {
		outputDir = asString(cfg["output_dir"])
	}
	if outputDir == "" {
		outputDir = a.cfg.DefaultOutDir
	}
	if serverBase == "" {
		serverBase = asString(cfg["server_base"])
	}
	if serverBase == "" {
		serverBase = a.cfg.defaultServerBase()
	}
	outRootDir := filepath.Clean(outputDir)

	var allLibs []map[string]any
	for _, info := range a.cfg.ListLibraries() {
		id, _ := info["id"].(string)
		if id == "" {
			continue
		}
		lib, err := a.cfg.LoadLib(id)
		if err == nil {
			allLibs = append(allLibs, lib)
		}
	}

	expectedStrm, expectedSide, strmMap, sideMap := collectExpected(allLibs, outputDir, ds)
	result := a.syncCore(expectedStrm, expectedSide, strmMap, sideMap, outputDir, serverBase, ds, false, "all", outRootDir)

	log.Printf("[同步] 完成: 删除 STRM %d 个 / 附属文件 %d 个, 生成 STRM %d 个 / 附属文件 %d 个, 错误 %d 个",
		len(result["deleted_strm"].([]string)),
		len(result["deleted_subs"].([]string)),
		len(result["created_strm"].([]string)),
		len(result["created_subs"].([]string)),
		len(result["errors"].([]string)))
	if errs := result["errors"].([]string); len(errs) > 0 {
		for _, e := range errs {
			log.Printf("[同步] 错误: %s", e)
		}
	}

	return map[string]any{
		"libraries":            result,
		"total_expected_strm":  result["expected_strm"],
		"total_existing_strm":  result["existing_strm"],
		"total_to_delete_strm": result["to_delete_strm"],
		"total_to_create_strm": result["to_create_strm"],
		"total_deleted_strm":   len(result["deleted_strm"].([]string)),
		"total_created_strm":   len(result["created_strm"].([]string)),
		"total_deleted_subs":   len(result["deleted_subs"].([]string)),
		"total_created_subs":   len(result["created_subs"].([]string)),
		"total_errors":         len(result["errors"].([]string)),
	}
}

// ==================== MD5(etag) 去重 ====================

// dedupLibrary: 单库内按 etag 完全一致去重，返回待删清单（不执行删除）
func (a *App) dedupLibrary(lib map[string]any) map[string]any {
	libID, _ := lib["id"].(string)
	libName, _ := lib["name"].(string)
	files, _ := lib["files"].([]any)

	videoByEtag := map[string][]map[string]any{}
	subByEtag := map[string][]map[string]any{}
	for _, f := range files {
		fm, ok := f.(map[string]any)
		if !ok {
			continue
		}
		etag := strings.TrimSpace(asString(fm["etag"]))
		fpath := asString(fm["path"])
		if etag == "" || fpath == "" {
			continue
		}
		ext := strings.ToLower(filepath.Ext(fpath))
		if VIDEO_EXTS[ext] {
			videoByEtag[etag] = append(videoByEtag[etag], fm)
		} else if SUBTITLE_EXTS[ext] {
			subByEtag[etag] = append(subByEtag[etag], fm)
		}
	}

	buildGroups := func(groups map[string][]map[string]any) []map[string]any {
		out := []map[string]any{}
		for etag, items := range groups {
			if len(items) < 2 {
				continue
			}
			filesOut := []map[string]any{}
			for _, f := range items {
				filesOut = append(filesOut, map[string]any{
					"idx":  firstInt64(f, "idx"),
					"path": asString(f["path"]),
					"size": firstInt64(f, "size"),
				})
			}
			out = append(out, map[string]any{
				"etag":  etag,
				"size":  firstInt64(items[0], "size"),
				"count": len(items),
				"files": filesOut,
			})
		}
		return out
	}

	videoGroups := buildGroups(videoByEtag)
	subGroups := buildGroups(subByEtag)
	videoCount := 0
	for _, g := range videoGroups {
		videoCount += int(firstInt64(g, "count"))
	}
	subCount := 0
	for _, g := range subGroups {
		subCount += int(firstInt64(g, "count"))
	}

	return map[string]any{
		"lib_id":          libID,
		"lib_name":        libName,
		"video_groups":    videoGroups,
		"sub_groups":      subGroups,
		"video_dup_count": videoCount,
		"sub_dup_count":   subCount,
	}
}

// dedupScanAll: 扫描全部库，返回去重清单（不执行删除）
func (a *App) dedupScanAll() map[string]any {
	var results []map[string]any
	totalVideo := 0
	totalSub := 0
	for _, info := range a.cfg.ListLibraries() {
		id, _ := info["id"].(string)
		if id == "" {
			continue
		}
		lib, err := a.cfg.LoadLib(id)
		if err != nil {
			continue
		}
		r := a.dedupLibrary(lib)
		if len(r["video_groups"].([]map[string]any)) > 0 || len(r["sub_groups"].([]map[string]any)) > 0 {
			results = append(results, r)
		}
		totalVideo += int(firstInt64(r, "video_dup_count"))
		totalSub += int(firstInt64(r, "sub_dup_count"))
	}
	return map[string]any{
		"libraries":       results,
		"total_video_dup": totalVideo,
		"total_sub_dup":   totalSub,
		"total_dup":       totalVideo + totalSub,
	}
}

// dedupApply: 对指定库执行删除：从库 JSON 移除勾选的 path，删除前自动备份
func (a *App) dedupApply(libID string, deletePaths []string) map[string]any {
	lib, err := a.cfg.LoadLib(libID)
	if err != nil {
		return map[string]any{"ok": false, "error": "library not found"}
	}
	files, _ := lib["files"].([]any)
	delSet := map[string]bool{}
	for _, p := range deletePaths {
		delSet[p] = true
	}
	keep := []any{}
	for _, f := range files {
		fm, ok := f.(map[string]any)
		if !ok {
			continue
		}
		if !delSet[asString(fm["path"])] {
			keep = append(keep, fm)
		}
	}
	removed := len(files) - len(keep)
	if removed == 0 {
		return map[string]any{"ok": true, "lib_id": libID, "removed": 0, "message": "无删除项"}
	}

	// 备份原库 JSON
	p := a.cfg.LibPath(libID)
	bak := fmt.Sprintf("%s.bak.%d", p, time.Now().Unix())
	if b, err := os.ReadFile(p); err == nil {
		os.WriteFile(bak, b, 0o644)
	}

	// 重建 idx 并写回
	newFiles := []any{}
	for i, f := range keep {
		fm, _ := f.(map[string]any)
		nf := map[string]any{}
		for k, v := range fm {
			nf[k] = v
		}
		nf["idx"] = i
		newFiles = append(newFiles, nf)
	}
	lib["files"] = newFiles
	a.cfg.writeLibraryFile(p, lib)
	return map[string]any{"ok": true, "lib_id": libID, "removed": removed, "message": fmt.Sprintf("已删除 %d 个重复文件", removed)}
}
